### PR TITLE
`ui.Chat()` now correctly handles new `ollama.chat()` return value introduced in ollama 0.4

### PR DESCRIPTION
**Pull request by cpsievert**
Ollama 0.4 [changed the return type of `ollama.chat()`](https://github.com/ollama/ollama-python/pull/276) (`ChatResponse`) from a `TypedDict` to a `pydantic.BaseModel`. As a result, passing that return value directly to `.append_message()` or `.append_message_stream()` doesn't work:

This PR fixes the issue, adds a unit test, and fixes another failing unit test.

```python
# ------------------------------------------------------------------------------------
# A basic Shiny Chat example powered by Ollama.
# To run it, you'll need an Ollama server running locally.
# To download and run the server, see https://github.com/ollama/ollama
# To install the Ollama Python client, see https://github.com/ollama/ollama-python
# ------------------------------------------------------------------------------------

import ollama

from shiny.express import ui

# Set some Shiny page options
ui.page_opts(
    title="Hello Ollama Chat",
    fillable=True,
    fillable_mobile=True,
)

# Create and display empty chat
chat = ui.Chat(id="chat")
chat.ui()


# Define a callback to run when the user submits a message
@chat.on_user_submit
async def _():
    # Get messages currently in the chat
    messages = chat.messages(format="ollama")
    # Create a response message stream
    # Assumes you've run `ollama run llama3` to start the server
    response = ollama.chat(
        model="llama3",
        messages=messages,
        stream=True,
    )
    # Append the response stream into the chat
    await chat.append_message_stream(response)
```

Screenshot 2024-11-26 at 11 12 33 AM

### CODE REVIEW

File: shiny/ui/_chat_tokenizer.py [hunk 1]
```diff
@@ -45,12 +45,23 @@ def encode(
...
+            "package (`pip install tokenizers`). "
+        )
+    except Exception as e:
+        raise ValueError(
+            "Failed to download a default tokenizer. "
+            "A tokenizer is required to impose `token_limits` on `chat.messages()`. "
+            "Try downloading a different tokenizer using "
+            "`tokenizers.Tokenizer.from_pretrained()`. "
+            f"Error: {e}"
+        ) from e
```
Comment (cpsievert) [inline shiny/ui/_chat_tokenizer.py:67]: This change is orthogonal to the main fix of this PR, but it's a good idea, and I ran into because [bert-base-cased](https://huggingface.co/google-bert/bert-large-cased) temporarily went offline for a couple of the test runs.

File: shiny/ui/_chat_tokenizer.py [hunk 2]
```diff
@@ -45,12 +45,23 @@ def encode(
...
     try:
         from tokenizers import Tokenizer
 
         return Tokenizer.from_pretrained("bert-base-cased")  # type: ignore
-    except Exception:
-        pass
-
-    return None
+    except ImportError:
+        raise ValueError(
```
Comment (gadenbuie) [inline shiny/ui/_chat_tokenizer.py:54]: Seems like `ImportError()` would be a good fit for our error too

File: shiny/ui/_chat_normalize.py [hunk 3]
```diff
@@ -231,11 +231,19 @@ def normalize_chunk(self, chunk: "dict[str, Any]") -> ChatMessage:
...
+            from ollama import ChatResponse
+
+            # Ollama<0.4 used TypedDict (now it uses pydantic)
+            # https://github.com/ollama/ollama-python/pull/276
+            if isinstance(ChatResponse, dict):
+                return "message" in message and super().can_normalize(
+                    message["message"]
+                )
+            else:
+                return isinstance(message, ChatResponse)
```
Comment (gadenbuie) [inline shiny/ui/_chat_normalize.py:244]: Is some of the context here that you have a message normalizer for Pydantic models?

Comment (cpsievert) [reply shiny/ui/_chat_normalize.py:244]: No, `DictNormalizer` will work for either case since they define `__getitem__()` on the pydantic model. I suppose that is a weird/subtle thing that requires extra context, and it'd be nice to take advantage of stronger pydantic typing, but I opted for the minimal change (especially if we're going to support older versions)

File: shiny/ui/_chat_tokenizer.py (hunk 1 above)
Comment (gadenbuie) [reply shiny/ui/_chat_tokenizer.py]: Seems like the second message is missing some instruction about how to choose a specific tokenizer that was in the original message. I also think it'd be nice to include how to turn off the need for a tokenizer, but that might be obvious from the code you'd write to get here.
